snapcraft/models/project.py

```diff
@@ -426,6 +445,54 @@ def _validate_default_provider(cls, default_provider):
         return default_provider
 
 
+class Platform(pydantic.BaseModel):
+    """Snapcraft project platform definition."""
+
+    build_on: list[SnapArch] | None = pydantic.Field(min_items=1, unique_items=True)
+    build_for: list[SnapArch] | None = pydantic.Field(min_items=1, unique_items=True)
+
+    class Config:  # pylint: disable=too-few-public-methods
+        """Pydantic model configuration."""
+
+        allow_population_by_field_name = True
+        alias_generator: Callable[[str], str] = lambda s: s.replace(  # noqa: E731
+            "_", "-"
+        )
+
+    @pydantic.validator("build_for", pre=True)
+    @classmethod
+    def _vectorise_build_for(cls, val: str | list[str]) -> list[str]:
+        """Vectorise target architecture if needed."""
+        if isinstance(val, str):
+            val = [val]
+        return val
+
+    @pydantic.root_validator(skip_on_failure=True)
+    @classmethod
+    def _validate_platform_set(cls, values: Mapping[str, Any]) -> Mapping[str, Any]:
+        """Validate the build_on build_for combination."""
+        build_for: list[str] = values["build_for"] if values.get("build_for") else []
+        build_on: list[str] = values["build_on"] if values.get("build_on") else []
+
+        # We can only build for 1 arch at the moment
+        if len(build_for) > 1:
+            raise CraftValidationError(
+                str(
+                    f"Trying to build a rock for {build_for} "
+                    "but multiple target architectures are not "
+                    "currently supported. Please specify only 1 value."
+                )
+            )
+
+        # If build_for is provided, then build_on must also be
+        if not build_on and build_for:
```

Comment:
Does this mean that this would be a valid platform:

```
platforms:
  amd64:
    build-on: [amd64, arm64]
```

Reply:
Correct. In this scenario, `build-for` would be the platform name.

Reply:
Good to know! I don't think charmcraft supports that so I should probably modify charmcraft.